### ISSUE TITLE
Go to definition cannot work as expected

### ISSUE THREAD

**Issue by davisduan** [Microsoft/TypeScript#15218]
Here is the case:
```
var x = {
    "abc": function () {
        return 1;
    },
    "def": function () {
        this.abc(); //cannot go to definition
    }
};
```
This function works well in sublime

**Comment by mhegazy** [comment]
The issue is not go-to-def. the issue is that the type of  `this` is `any` in an object literal by default. the rational here is object literals are passed around as property bags, and can be accessed with different `this` arguments.

If you are running with TS 2.3 or later, we have a [change](https://github.com/Microsoft/TypeScript/pull/14141) only under `--noImplicitThis` that allows for a stricter type of `this` in object literals among other things. go to def should work as expected under this flag.

**Comment by davisduan** [comment]
@mhegazy Thanks for your explanation. It works after I added the parameter. But I meet the same problem when I put the object as parameter:
```
Child.extend("Paraent", {
    "abc": function () {
        return 1;
    },
    "def": function () {
        this.abc(); //cannot go to definition
    }
});
```

**Comment by mhegazy** [comment]
how is `Parent.extend` defined?

**Comment by davisduan** [comment]
I don't think it's quite related to the function definition. But  you can take jQuery extend function as reference.
```
jQuery.extend({
    "abc": function () {
        return 1;
    },
    "def": function () {
        this.abc(); //cannot go to definition - the type of this is any
    }
});
```

**Comment by mhegazy** [comment]
`JQuery.extend` is defined with `target : any`; this pushes a contextual type of `any` for the function, and thus the `this` parameter.

if it was defined using generics instead:
```ts
declare function extend<T>(target: T): any;
```

you would get your `this` type correctly under `--noImplicitThis`.

You can take this one step further and make the type of `this` the combination of all the object literal properties using `ThisType<T>`, so:

```ts

declare function extend<T, U, V>(target: T & ThisType<T & U & V>, object1: U & ThisType<T & U & V>, object2: V & ThisType<T & U & V>): any;

extend({
        "abc": function () {
            return 1;
        },
        "def": function () {
            this.abc(); // OK
        }
    },
    {
        "another": function () {
            this.def(); // OK
         }
    },
    {
        "yetAnother": function () {
            this.another(); // OK
         }
    });
```

you can find more info about the use of `ThisType` in https://github.com/Microsoft/TypeScript/pull/14141

**Comment by davisduan** [comment]
Actually, I just want to use Visual Studio Code to develop in JavaScript, which is is using TypeScript as JS compiler. But I cannot use "declare function ..." in JS code. 
Seems it too complex for me to achieve the "Go to definition" function works like sublime.
Anyway, thank you very much. @mhegazy

**Comment by mhegazy** [comment]
Where is extends comming from? VSCode gets your declarations form an @types package, so updating it should do the trick without you having to change your code.

**Comment by davisduan** [comment]
Extends is one of functions in my JS library. Here is only a example. I don't think I can add all the functions declaration manually.

**Comment by mhegazy** [comment]
This issue seems to have been already addressed or is unactionable at the moment. I am auto-closing it for now.
